Copy bundles from Lock file

DennisDenuto opened the pull request:
This is a draft (so so draft it could be a beer 🍺) PR to increase visibility on changes being made. As well as keep track of any design decisions being made as work is iterated on.

- Fixes e2e test. (e2e were failing when run against dockerhub)
- Always copy recursively from lockfile (does not yet honor the recursive flag)
- Refactor helpers fake_registry.go <- spins up a v2 complaint registry. This results in a lot less mock code that has to be maintained.


EDIT: No longer a draft PR :-)

DennisDenuto committed:
Fix e2e test

- when run against a public registry, this e2e would fail (403) due to the
repository being truncated out when pushing 'test' images and bundles.

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Copy from bundle lock file

- Always recursively copies bundles from lock file, without honoring the
recursive feature flag

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor helpers fake registry

- uses a oci v2 compliant registry running as a http test server

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Remove test

- If blob mounting were to break, performance tests will catch that regression

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto commented:
Task List:

- [x] Maintain behavior without --experimental-recursive-bundle on simple bundles using a bundle lock file
- [x] Fail without --experimental-recursive-bundle on recursive bundle bundles using a bundle lock file
- [x]  Fail without --experimental-recursive-bundle on recursive bundle bundles using a bundle name (-b)
- [x] Fails when try to copy bundle using an image lock file
- [x] reduce verbosity of fake registry
- [x] replace 'duplicate' fake_images_metadata_builder_test for helpers.fake_registry

DennisDenuto committed:
Backfill tests for copying nested bundles

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Provide error if copying a nested bundle without the --recursive flag

Co-authored-by: Dennis Leon <leonde@vmware.com>
Co-authored-by: Garrett Cheadle <gcheadle@vmware.com>

DennisDenuto commented:
While implementing Scenario 3 from issue https://github.com/vmware-tanzu/carvel-imgpkg/issues/86, I noticed that issue https://github.com/vmware-tanzu/carvel-imgpkg/issues/85 hadn't implemented the same error case scenario. So as part of this PR, in order to maintain consistency between a bundle lock file and a bundle name, I will also implement scenario 3 for issue #85 .

This has performance implications. In order to know whether a bundle contains a nested bundle, a request needs to be made to the registry.

DennisDenuto committed:
Perf improvement

- if the experimental recursive feature flag is not enabled, then only
check the root bundle's images lock file to detect if there are any
nested bundles.

Co-authored-by: Dennis Leon <leonde@vmware.com>
Co-authored-by: Garrett Cheadle <gcheadle@vmware.com>

DennisDenuto committed:
Provide error if copying a nested bundle

- This covers the case when a bundle was specified using a bundle name
(-b) without setting the experimental recursive feature flag

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Refactor

- extract getting a bundle (that shouldn't contain a nested bundle)
images lock localized into a function shared by specifying a bundle via
a lock file and the bundle (-b) flag

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
UX improvement

- when providing an imageslock file that references a bundle, provide a meaningful error message to the user

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Test refactor

- make registry server less verbose. Discards logs. It was previously
printing every request made to the registry.
- Remove duplicate fake registry and use the one in the helpers package
- Remove duplicate test code

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/imgpkg/cmd/copy_repo_src.go:
```diff
@@ -175,6 +187,26 @@ func (o CopyRepoSrc) getSourceImages() (*ctlimgset.UnprocessedImageRefs, error)
+		}
+		return lockconfig.ImagesLock{}, err
+	}
+	for _, i := range imagesLock.Images {
+		isBundle, err := ctlbundle.NewBundle(i.Image, o.registry).IsBundle()
+		if err != nil {
+			return lockconfig.ImagesLock{}, err
+		}
+		if isBundle {
+			return lockconfig.ImagesLock{}, fmt.Errorf("This bundle contains bundles, in order to copy please use the --experimental-recursive-bundle flag to copy nested bundles")
+		}
+	}
```
no need to check this here since push is verifying this: https://github.com/vmware-tanzu/carvel-imgpkg/blob/f0aadec0ec3e58bf4927be4f7ad87bd112a585dc/pkg/imgpkg/bundle/contents.go#L89-L92 (was already prepped for recursive bundles with a allowInnerBundles bool.

DennisDenuto replied:
Hmmm i'm not following / missing a nuance here...

This check happens during a `copy`. Specifically they are trying to copy a bundle that contains nested bundles, but they haven't provided the feature flag. In this case we want to provide a meaningful error message to the user, teaching them about this flag.

From the issue, it is implementing `Scenario 3: Fail without --experimental-recursive-bundle on recursive bundle bundles using a bundle lock file`

cppforlife replied:
if we remove this function before release then this doesnt matter.

in case we keep this function in the next release then this affects performance of regular users that use imgpkg because we would be iterating over each image and fetching its config to check if something is a bundle. this flag was made to ensure that we do not "ship" recursive bundle code too early afaik, not with an intent that we would be encouraging ppl to use it (since it's not ready for consumption). (in other words this code favors experimental users instead of regular users).

(code i linked above prevents users from making bundles with nested bundles if flag is not used, hence we know that user knows about the flag).

DennisDenuto replied:
> if we remove this function before release then this doesnt matter.

True.

I am going to assume we ship a release *with* the feature flag still in... Because, you know, an urgent security fix had to be applied...

In that case, the only risk I can think of is, if someone tries to copy a bundle (that contains a nested bundle), and they do *not* specify the feature flag (because to err is human), then the copy command will succeed (Even though the nested bundles haven't been copied)

Eh, The above concern is probably unlikely to happen in practice. So I'll remove the check and update the tests too to reflect this change in behavior.

DennisDenuto committed:
Improve performance

- When the recursive feature flag is not provided, stop checking if the bundle contained any sub bundles (in order to
return an error to the user). This check was making requests to the
registry. Instead, don't recursively copy the nested bundles.

For more info see discussion: https://github.com/vmware-tanzu/carvel-imgpkg/pull/104#discussion_r597302314

Authored-by: Dennis Leon <leonde@vmware.com>
